ddl/index.go

```diff
@@ -698,6 +700,380 @@ func (w *worker) doBackfillIndexTaskInTxn(t table.Table, txn kv.Transaction, col
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+
+	return startHandle, endHandle, nil
+}
+
+func closeAddIndexWorkers(workers []*addIndexWorker) {
+	for _, worker := range workers {
+		worker.close()
+	}
+}
+
+func (d *ddl) handleTaskResults(workers []*addIndexWorker, workingIdx int) (int64, int64, error) {
+	results := make([]*addIndexResult, 0, len(workers))
+	for i, worker := range workers {
+		if i > workingIdx {
+			break
+		}
+		result := <-worker.resultCh
+		results = append(results, result)
+	}
+
+	var firstErr error
+	var addedCount int64
+	var nextHandle int64
+	for i, result := range results {
+		if i == 0 {
+			// init nextHandle
+			nextHandle = result.originalTaskRange.startHandle
+		}
+		if firstErr == nil && result.err != nil {
+			firstErr = result.err
+			if errCancelledDDLJob.Equal(firstErr) {
+				nextHandle = result.nextHandle
+			}
+		}
+
+		// count it any way.
+		addedCount += int64(result.count)
+		if firstErr == nil {
+			nextHandle = result.nextHandle
+		}
+	}
+
+	return nextHandle, addedCount, firstErr
+}
+
+func (d *ddl) finishBatchTasks(startTime time.Time, startHandle int64, reorgInfo *reorgInfo, job *model.Job, workers []*addIndexWorker, workingIdx int) error {
+	addedCount := job.GetRowCount()
+	nextHandle, taskAddedCount, err := d.handleTaskResults(workers, workingIdx)
+	addedCount += taskAddedCount
+	elapsedTime := time.Since(startTime).Seconds()
+	if err == nil {
+		err = d.isReorgRunnable()
+	}
+
+	if err != nil {
+		// update the reorg handle that has been processed.
+		err1 := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
+			return errors.Trace(reorgInfo.UpdateHandle(txn, nextHandle))
+		})
+		log.Warnf("[ddl-reorg] total added index for %d rows, this task [%d,%d) add index for %d failed %v, take time %v, update handle err %v",
+			addedCount, startHandle, nextHandle, taskAddedCount, err, elapsedTime, err1)
+		return errors.Trace(err)
+	}
+
+	// nextHandle will be updated periodically in runReorgJob, so no need to update it here.
+	d.reorgCtx.setRowCountAndHandle(addedCount, nextHandle)
+	metrics.BatchAddIdxHistogram.Observe(elapsedTime)
+	log.Infof("[ddl-reorg] total added index for %d rows, this task [%d,%d) added index for %d rows, take time %v",
+		addedCount, startHandle, nextHandle, taskAddedCount, elapsedTime)
+	return nil
+}
+
+func (d *ddl) addTableIndexFromSplitRanges(t table.Table, indexInfo *model.IndexInfo, reorgInfo *reorgInfo, job *model.Job) error {
+	colFieldMap := makeupIndexColFieldMap(t, indexInfo)
+
+	workerCnt := defaultWorkers
+	workers := make([]*addIndexWorker, workerCnt)
+	for i := 0; i < workerCnt; i++ {
+		sessCtx := d.newContext()
+		index := tables.NewIndexWithBuffer(t.Meta(), indexInfo)
+		workers[i] = newAddIndexWoker(sessCtx, d, i, index, t, colFieldMap)
+		go workers[i].run()
+	}
+	defer closeAddIndexWorkers(workers)
+
+	kvRanges, err := d.splitTableRanges(t, reorgInfo.Handle)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	workerIdx := 0
```

Comment:
This variable makes the code complex.
How about we split the ranges into batch tasks first and then sequentially do each batch task?